### ISSUE TITLE
Coordination fails on dynamic/parallel branching for intermediate results

### ISSUE THREAD

**Issue by chennqqi** [s8sg/faas-flow#64]
https://github.com/s8sg/faas-flow/blob/77c988e2c5e46b48438fa60ce7119340276ac37c/context.go#L169

dataStore.Get return empty data.  while when i checked minio, this data exist.
this bug not trigger everytime.  like https://github.com/s8sg/faas-flow/pull/61

**Comment by chennqqi** [comment]
```
        // Recive data from a dynamic graph for each options
        time.Sleep(2 * time.Second)
        // Recive data from a dynamic graph for each options
        time.Sleep(2 * time.Second)
        for _, option := range options {

                // skip retriving data for current option
                if option == currentOption {
                        continue
                }

                key := fmt.Sprintf("%s--%s--%s",
                        option, pipeline.GetNodeExecutionUniqueId(currentNode), currentNode.GetUniqueId())
                idata := context.GetBytes(key)
```
I add two second time sleep. `context.GetBytes(key)` return corrected data.
maybe some sync cache problem.

**Comment by chennqqi** [comment]
I create 
https://github.com/chennqqi/faas-flow-redis-datastore for redis datastore
https://github.com/chennqqi/faas-flow-redis-stateestore for redis statestore
to debug easly

**Comment by s8sg** [comment]
@chennqqi thank you for this. 

It seems that both branch can executes at the same time, and it's possible if the 2nd one gets executed even before the earlier branch successfully store the result in `DB`.  
Its definitely sync issue. Nice Catch. 

I'm open for a solution. Distributed lock might work, is there anything we can do without changing interface of `StateStore`?

**Comment by chennqqi** [comment]
New version support of-watchdog. Thus all requests of the `flow node` are asyn.
Old version use classic watchdog, every request of node execution in the `flow node` run one by one.

**Comment by chennqqi** [comment]
https://github.com/s8sg/faas-flow/blob/77c988e2c5e46b48438fa60ce7119340276ac37c/template/faas-flow/handler.go#L825

```go
		inDegreeUpdatedCount, err := fhandler.IncrementCounter(key, 1)
		if err != nil {
			return []byte(""), fmt.Errorf("failed to update inDegree counter for node %s", currentNode.GetUniqueId())
		}

		// not all the branches finished executing
		if inDegreeUpdatedCount < len(options) {

			// Skip if dynamic node data forwarding is not disabled
			if currentNode.GetForwarder("dynamic") != nil {
				option := pipeline.CurrentDynamicOption[currentNode.GetUniqueId()]
				key = fmt.Sprintf("%s--%s--%s", option, pipeline.GetNodeExecutionUniqueId(currentNode), currentNode.GetUniqueId())

				err = context.Set(key, result)
				if err != nil {
					return nil, fmt.Errorf("failed to store branch result of dynamic node %s for option %s, error %v",
						currentNode.GetUniqueId(), option, err)
				}
				fmt.Printf("[Request `%s`] Intermidiate result from Branch to Dynamic Node %s for option %s stored as %s\n",
					fhandler.id, currentNode.GetUniqueId(), option, key)
			}

			fmt.Printf("[Request `%s`] Executing end of dynamic node %s, delayed as "+
				"completed indegree: %d/%d\n",
				fhandler.id, currentNode.GetUniqueId(), inDegreeUpdatedCount, len(options))
			// Return nil response making current iteration to stop
			return nil, nil
		}
```
I think should  `context.Set(key, result)` first then `IncrementCounter`. reason is:
all dynamic node of a dag/subdag run in differrent go routine in a same flow function instances.
`IncrementCounter` run first, other one go routine indegree match expect then run context.GetBytes.
if context.GetBytes before context.Set. will cause a painic.

**Comment by s8sg** [comment]
@chennqqi Thats one option, I though of it too. But if you see carefully we do not store the result in datastore for the last branch execution, we just put the result in the aggregator hash-map `subDataMap `. This way we save some time, cause lastBranch routine will perform the aggregator. 
```go
// Store the current data
subDataMap[currentOption] = result
.
.
.
// Recive data from a dynamic graph for each options
for _, option := range options {
	// skip retriving data for current option
	if option == currentOption {
		continue
	}
```
As increment counter is the only way to know which one is the last branch, we can either save the time and find other way to sync, or store data in dataStore even for last branch.

**Comment by chennqqi** [comment]
I understand. last dynamic branch run aggregator.
The question is how to solve sync problem, to avoid fetch results panic before aggregator.
Two ways:
1. two-step submit. add new counter for check if all `context.Set` done before `context.GetBytes`
2. store data to datastore before `IncrementCounter`

**Comment by s8sg** [comment]
Option 2 looks cleaner to me. Will you be interested to implement the change ?

**Comment by chennqqi** [comment]
I think i can try. but i'm  not confidence in solving this problem.

**Comment by s8sg** [comment]
@chennqqi you understand the problem well, good if you can give it a try. We can test whether it solved the problem or not after that

**Comment by chennqqi** [comment]
https://github.com/s8sg/faas-flow/pull/61#issuecomment-513624538
I found this problem maybe caused by consul. I use my code https://github.com/chennqqi/faas-flow-redis-statestore as statestore which made result perfect. while using consul as statestore, the result some times were correct, sometimes no data return.

I removed https://github.com/s8sg/faas-flow-consul-statestore/tree/master/vendor fold. because this fold would cause 'filename too long' error.

Conclusion, something wrong in `consulstatestore.Update` function
